### PR TITLE
Removing dimension from TensorView

### PR DESCRIPTION
**Pull request by szalpal**
Signed-off-by: Michał Szołucha <mszolucha@nvidia.com>

### CODE REVIEW

File: dali/kernels/test/tensor_view_test.cc [hunk 1]
```diff
@@ -218,5 +218,33 @@ TEST(TensorListView, uniform_list_shape) {
...
+  }
+}
+
+}  // namespace
+
+TEST(TensorViewTest, OustDimensionTest) {
+ using namespace std;
+ constexpr size_t kNDims = 4;
+ array<long int, kNDims> dims = {4,1,2,3};
+ vector<int> data(accumulate(dims.begin(), dims.end(), 1, multiplies<int>()), 0);
```
Comment (mzient) [inline dali/kernels/test/tensor_view_test.cc]:
```suggestion
 vector<int> data(volume(dims), 0);
```

File: dali/kernels/test/tensor_view_test.cc [hunk 2]
```diff
@@ -218,5 +218,33 @@ TEST(TensorListView, uniform_list_shape) {
...
+ {
+   int idx = 0;
+   for_each(data.begin(), data.end(), [&](int &val) { val += idx++; });
```
Comment (mzient) [inline dali/kernels/test/tensor_view_test.cc]:
```suggestion
   std::iota(data.begin(), data.end(), 0);
```

File: dali/kernels/test/tensor_view_test.cc [hunk 3]
```diff
@@ -218,5 +218,33 @@ TEST(TensorListView, uniform_list_shape) {
   EXPECT_EQ(infer2, dyn);
 }
 
+namespace {
+template<typename T, size_t N>
+void Verify(const T* data, std::array<long int, N> dims, int idx) {
```
Comment (mzient) [inline dali/kernels/test/tensor_view_test.cc]: Less generic name please - it's case-specific and should be named so.

File: dali/kernels/test/tensor_view_test.cc [hunk 4]
```diff
@@ -218,5 +218,33 @@ TEST(TensorListView, uniform_list_shape) {
...
+  auto f = std::accumulate(dims.begin() + 1, dims.end(), 1, std::multiplies<int>());
```
Comment (mzient) [inline dali/kernels/test/tensor_view_test.cc]:
```suggestion
  auto sub_vol = volume(dims.begin() + 1, dims.end());
```

File: dali/kernels/test/tensor_view_test.cc [hunk 5]
```diff
@@ -218,5 +218,33 @@ TEST(TensorListView, uniform_list_shape) {
...
+  for (int i = 0; i < f; i++) {
```
Comment (mzient) [inline dali/kernels/test/tensor_view_test.cc]:
```suggestion
  for (int i = 0; i < sub_vol; i++) {
```

File: dali/kernels/test/tensor_view_test.cc [hunk 6]
```diff
@@ -218,5 +218,33 @@ TEST(TensorListView, uniform_list_shape) {
...
+    EXPECT_EQ(idx * f + i, data[i]) << "Failed on idx: " + std::to_string(i);
```
Comment (mzient) [inline dali/kernels/test/tensor_view_test.cc]:
```suggestion
    EXPECT_EQ(idx * sub_vol + i, data[i]) << "Failed at idx: " << i;
```

File: dali/kernels/tensor_view.h [hunk 7]
```diff
@@ -414,6 +414,33 @@ TensorListView<StorageGPU, T, ndim> make_tensor_list_gpu(T *data, TensorListShap
   return { data, std::move(shape) };
 }
 
+
+/**
```
Comment (mzient) [inline dali/kernels/tensor_view.h]: The whole header uses /// doxygen style. Please be consistent.

File: dali/kernels/test/tensor_view_test.cc (hunk 3 above)
Comment (szalpal) [reply dali/kernels/test/tensor_view_test.cc]: done

File: dali/kernels/test/tensor_view_test.cc (hunk 4 above)
Comment (szalpal) [reply dali/kernels/test/tensor_view_test.cc]: done

File: dali/kernels/test/tensor_view_test.cc (hunk 5 above)
Comment (szalpal) [reply dali/kernels/test/tensor_view_test.cc]: done

File: dali/kernels/test/tensor_view_test.cc (hunk 6 above)
Comment (szalpal) [reply dali/kernels/test/tensor_view_test.cc]: done

File: dali/kernels/test/tensor_view_test.cc (hunk 1 above)
Comment (szalpal) [reply dali/kernels/test/tensor_view_test.cc]: done

File: dali/kernels/test/tensor_view_test.cc (hunk 2 above)
Comment (szalpal) [reply dali/kernels/test/tensor_view_test.cc]: done

File: dali/kernels/tensor_view.h (hunk 7 above)
Comment (szalpal) [reply dali/kernels/tensor_view.h]: This header is inconsistent with the rest of our code base. I want to be consistent with code base

Comment (mzient) [reply dali/kernels/tensor_view.h]: I think that file-scope consistency is more important than global one. Also, almost everything in dali/kernels uses this style.

File: dali/kernels/test/tensor_view_test.cc [hunk 8]
```diff
@@ -218,5 +230,31 @@ TEST(TensorListView, uniform_list_shape) {
...
+
+TEST(TensorViewTest, SubtensorTest) {
+  using namespace std;  // NOLINT
+  constexpr size_t kNDims = 4;
+  array<int64_t, kNDims> dims = {4, 1, 2, 3};
+  vector<int> data(volume(dims), 0);
+  iota(data.begin(), data.end(), 0);
+  auto tv = make_tensor_cpu<kNDims>(data.data(), dims);
+  for (int i = 0; i < dims[0]; i++) {
+    auto ret = subtensor(tv, static_cast<size_t>(i));
```
Comment (mzient) [inline dali/kernels/test/tensor_view_test.cc]: Why this static_cast here?

File: dali/kernels/tensor_view.h [hunk 9]
```diff
@@ -414,6 +414,32 @@ TensorListView<StorageGPU, T, ndim> make_tensor_list_gpu(T *data, TensorListShap
...
+ * tv.data = [[1, 2, 3], [4, 5, 6]]       (shape: [2, 3])
+ * oust_dimension(tv, 1) -> [4, 5, 6]     (shape: [3])
+ *
+ * @param source Source TensorView
+ * @param idx Index inside dimension, along which data is extracted
+ * @return TensorView with reduced dimensionality
+ */
+template<typename StorageBackend, typename DataType, int ndims>
+TensorView<StorageBackend, DataType, ndims - 1>
+subtensor(TensorView<StorageBackend, DataType, ndims> source, size_t idx) {
```
Comment (mzient) [inline dali/kernels/tensor_view.h]: use int64_t for index - that's what we use almost everywhere, this function shouldn't be different.

File: dali/kernels/tensor_view.h [hunk 10]
```diff
@@ -414,6 +414,32 @@ TensorListView<StorageGPU, T, ndim> make_tensor_list_gpu(T *data, TensorListShap
...
+ * Example:
+ * tv.data = [[1, 2, 3], [4, 5, 6]]       (shape: [2, 3])
+ * oust_dimension(tv, 1) -> [4, 5, 6]     (shape: [3])
+ *
+ * @param source Source TensorView
+ * @param idx Index inside dimension, along which data is extracted
+ * @return TensorView with reduced dimensionality
+ */
+template<typename StorageBackend, typename DataType, int ndims>
+TensorView<StorageBackend, DataType, ndims - 1>
```
Comment (mzient) [inline dali/kernels/tensor_view.h]:
```suggestion
TensorView<StorageBackend, DataType, out_ndim>
```

File: dali/kernels/tensor_view.h [hunk 11]
```diff
@@ -414,6 +414,32 @@ TensorListView<StorageGPU, T, ndim> make_tensor_list_gpu(T *data, TensorListShap
...
+ *
+ * Example:
+ * tv.data = [[1, 2, 3], [4, 5, 6]]       (shape: [2, 3])
+ * oust_dimension(tv, 1) -> [4, 5, 6]     (shape: [3])
+ *
+ * @param source Source TensorView
+ * @param idx Index inside dimension, along which data is extracted
+ * @return TensorView with reduced dimensionality
+ */
+template<typename StorageBackend, typename DataType, int ndims>
```
Comment (mzient) [inline dali/kernels/tensor_view.h]:
```suggestion
template<int out_ndim_ = InferDimensions, typename StorageBackend, typename DataType, int ndim,
    int out_ndim = (out_ndim_ == InferDimensions ? ndim == DynamicDimensions ? DynamicDimensions : ndim-1 : out_ndim_>
```
This allows user to make a slice with known static number of dimensions from a tensor with dynamic number of dimensions.

File: dali/kernels/tensor_view.h (hunk 9 above)
Comment (mzient) [inline dali/kernels/tensor_view.h]: Add an overload for TensorView with DynamicDimensions.

Comment (szalpal) [reply dali/kernels/tensor_view.h]: Why? It's not like idx (or pos, whatever) can be <0...

File: dali/kernels/tensor_view.h [hunk 12]
```diff
@@ -414,6 +414,39 @@ TensorListView<StorageGPU, T, ndim> make_tensor_list_gpu(T *data, TensorListShap
...
+}
+
+
+/// @brief Overload for Dynamic TensorView
+template<typename StorageBackend, typename DataType>
+TensorView<StorageBackend, DataType, DynamicDimensions>
+subtensor(TensorView<StorageBackend, DataType, DynamicDimensions> source, int64_t pos) {
+  TensorShape<DynamicDimensions> shape = source.shape.last(source.dim() - 1);
+  DataType *data = source.data + pos * volume(shape);
+  return make_tensor<StorageBackend>(data, shape);
```
Comment (mzient) [inline dali/kernels/tensor_view.h]:
```suggestion
  return make_tensor<StorageBackend>(data, std::move(shape));
```

File: dali/kernels/tensor_view.h [hunk 13]
```diff
@@ -414,6 +414,39 @@ TensorListView<StorageGPU, T, ndim> make_tensor_list_gpu(T *data, TensorListShap
...
+}
+
+
+/// @brief Overload for Dynamic TensorView
+template<typename StorageBackend, typename DataType>
+TensorView<StorageBackend, DataType, DynamicDimensions>
+subtensor(TensorView<StorageBackend, DataType, DynamicDimensions> source, int64_t pos) {
+  TensorShape<DynamicDimensions> shape = source.shape.last(source.dim() - 1);
```
Comment (mzient) [inline dali/kernels/tensor_view.h]:
```suggestion
  auto shape = source.shape.last(source.dim() - 1);
```

File: dali/kernels/test/tensor_view_test.cc [hunk 14]
```diff
@@ -218,5 +230,45 @@ TEST(TensorListView, uniform_list_shape) {
...
+  for (int i = 0; i < dims[0]; i++) {
+    auto ret = subtensor(tv, i);
+    VerifySubtensor(ret.data, dims, i);
+  }
+}
+
+
+TEST(TensorViewTest, DynamicSubtensorTest) {
+  using namespace std;  // NOLINT
+  vector<int64_t> dims = {4, 1, 2, 3};
```
Comment (mzient) [inline dali/kernels/test/tensor_view_test.cc]: A suggestion: the first dimension of the sub-tensor should be >1 to be sure that the subtensor volume is calculated properly (nothing skipped).

**Comment by szalpal** [comment]
#671921

**Pull request merged**